### DISCUSSION TITLE
Why does TypeScript.tmLanguage clone only tokenize strings?

### DISCUSSION THREAD

**Discussion by martypdx** [microsoft/vscode-discussions#1008] [Extension Development QnA] [answered]
I thought I was starting to get how language and grammar work, until I tried the experiment below. The goal was to examine the behavior of the typescript grammar without getting any of the "built-in" behavior associated with `.ts` files.

Here is what I did:

1. Setup new vscode language project
2. Copied the `TypeScript.tmLanguage` file from [Typescript-tmLanguage](https://github.com/Microsoft/TypeScript-TmLanguage) with the name `Testy.tmLanguage.json` and making these changes: 
    1. replace all occurrences of `.ts` with `.testy`
    2. remove the frontmatter info and id and replace with:
    ```json
    "name": "TestScript",
    "scopeName": "source.testy",
    "fileTypes": [
      "testy"
    ], 
    ```
3. Setup `package.json` as follows:
```json
    "contributes": {
        "languages": [
            {
                "id": "testy",
                "extensions": [
                    ".testy"
                ],
                "aliases": [
                    "Testy"
                ]
            }
        ],
        "grammars": [
            {
                "path": "./syntaxes/Testy.tmLanguage.json",
                "scopeName": "source.testy",
                "language": "testy"
            }
        ]
    },
```

F5 and added a `test.testy` file. My expectation was that I'd get basic typescript formatting, instead I got:

![image](https://github.com/microsoft/vscode-discussions/assets/478864/a7b6acf7-f2dd-47d2-9a84-a5ba200ca7aa)

It tokenizes strings but nothing else. And the grammar is loaded because we can see the scopes (and if I comment out `-include: '#strings'` it stops tokenizing those). Here's the start of the grammar file which shouldn't have change in structure:

![image](https://github.com/microsoft/vscode-discussions/assets/478864/de0508d9-09d1-4777-b877-b60d70bebc20)

What am I missing here?

**Comment by martypdx** [comment, marked as answer]
**Update:** if I use `TypeScriptReact.tmLanguage` it works as expected:

![image](https://github.com/microsoft/vscode-discussions/assets/478864/11c6ad5e-27a4-41c8-8e43-34baa7e97775)

**Comment by martypdx** [comment]
Would be great to know why these two files differ. 

I suspect what may be going on is that the included jsx patterns are bringing in the deeper patterns for matches. TypeScriptReact, JavaScript, and JavaScriptReact all include jsx. TypeScript is the only one of the four that doesn't have jsx. And it probably would not be deployed be default on its own without at least one of the other variants.

**Reply by RedCMD** [reply]
Judging based off the `"variables"` key in the picture
![image](https://github.com/microsoft/vscode-discussions/assets/33529441/f081ac78-b5d1-49d9-bf53-31fb2c12f848)
I assume you've converted the YAML file?
[TypeScript.YAML-tmLanguage](https://github.com/microsoft/TypeScript-TmLanguage/blob/master/TypeScript.YAML-tmLanguage)

they do a bunch of preprocessing with replacing variables in the regex's like `{{constantIdentifier}}` with the values in `"variables"`
before converting it to the JSON file

that would be my assumption without knowing exactly what files you have
